sem.h

```diff
@@ -17,12 +18,21 @@ sem_init(sem_t *sem, int pshared, int value)
     sem->value = value;
 }
 
+void
+exit_func(void *arg)
+{
+    sem_t *sem = (sem_t *) arg;
+    pthread_mutex_unlock(&sem->value_mutex);
+}
+
 void
 sem_wait(sem_t *sem)
 {
     pthread_mutex_lock(&sem->value_mutex);
+    pthread_cleanup_push(&exit_func, (void *) sem);
```

Comment:
Could be reduced to 
`pthread_cleanup_push ((void (*) (void*))pthread_mutex_unlock, &sem->value_mutex);`